ddtrace/contrib/molten/patch.py

```diff
@@ -0,0 +1,143 @@
+import inspect
+
+import wrapt
+
+import molten
+from ddtrace import Pin, config
+from ddtrace.utils.formats import get_env
+from ddtrace.utils.importlib import func_name
+
+from ...ext import AppTypes
+from ...ext import http
+
+# Configure default configuration
+config._add('molten', dict(
+    service_name=get_env('molten', 'service_name', 'molten'),
+    app='molten',
+    app_type=AppTypes.web,
+    distributed_tracing_enabled=False,
+))
+
+def patch():
+    """Patch the instrumented methods
+    """
+    if getattr(molten, '_datadog_patch', False):
+        return
+    setattr(molten, '_datadog_patch', True)
+
+    pin = Pin(
+        service=config.molten['service_name'],
+        app=config.molten['app'],
+        app_type=config.molten['app_type'],
+    )
+
+    # add pin to module since many classes use __slots__
+    pin.onto(molten)
+
+    _w = wrapt.wrap_function_wrapper
+    _w('molten', 'App.__init__', patch_app_init)
+    _w('molten', 'App.__call__', patch_app_call)
+    _w('molten', 'Router.add_route', patch_add_route)
+
+def trace_func(resource):
+    @wrapt.function_wrapper
+    def _trace_func(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        with pin.tracer.trace(func_name(wrapped), service=pin.service, resource=resource):
+            return wrapped(*args, **kwargs)
+
+    return _trace_func
+
+def trace_middleware(middleware):
+    @wrapt.function_wrapper
+    def _trace_middleware(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        name = None
+        if inspect.isfunction(wrapped):
+            name = wrapped.__name__
+        else:
+            name = type(wrapped).__name__
+
+        with pin.tracer.trace(name, service=pin.service):
+            return trace_func(name)(wrapped(*args, **kwargs))
+
+    return _trace_middleware(middleware)
+
+def patch_start_response(start_response):
+    @wrapt.function_wrapper
+    def _start_response(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+        span = pin.tracer.current_root_span()
+        status, headers, exc_info = args
+        status_code, _, _ = status.partition(' ')
+        span.set_tag(http.STATUS_CODE, status_code)
+        return wrapped(*args, **kwargs)
+
+    return _start_response(start_response)
+
+def patch_add_route(wrapped, instance, args, kwargs):
+    def _wrap(route_like, prefix="", namespace=None):
+        # avoid patching non-Route, e.g. Include
+        if not isinstance(route_like, molten.Route):
+            return wrapped(*args, **kwargs)
+
+        resource = '{} {}'.format(
+            route_like.method,
```

Comment:
How do you define a route for multiple methods?

e.g. A route that allows both `GET` and `POST`?

We might need to wait until we actually have the request to know which method is being requested by the user.

If it is a 1:1 relationship between method and route, then no changes are needed here.

Reply:
Talked offline, it is a 1:1 relationship.